src/workerd/jsg/modules-new-test.c++

```diff
@@ -0,0 +1,1889 @@
+    ResolveObserverImpl observer;
+    CompilationObserver compilationObserver;
+
+    ModuleBundle::BundleBuilder bundleBuilder;
+
+    // A recursive import with an ESM works just fine...
+    auto foo = kj::str("import foo from 'foo'; export default 123;");
+    bundleBuilder.addEsmModule("foo", foo.slice(0, foo.size()).attach(kj::mv(foo)));
+
+    // A CommonJS-style module, however, does not allow recursive evaluation.
+    bundleBuilder.addSyntheticModule("bar",
+        Module::newCjsStyleModuleHandler<TestType, TestIsolate_TypeWrapper>(
+            kj::str("require('bar')"),
+            kj::str("bar")));
+
+    auto registry = ModuleRegistry::Builder(observer).add(bundleBuilder.finish()).finish();
+
+    registry->attachToIsolate(js, compilationObserver);
+
+    auto val1 = ModuleRegistry::resolve(js, "file:///foo", "default"_kjc);
+    KJ_ASSERT(val1.isNumber());
+
+    js.tryCatch([&] {
+      ModuleRegistry::resolve(js, "file:///bar", "default"_kjc);
+    }, [&](Value exception) {
+      auto str = kj::str(exception.getHandle(js));
+      KJ_ASSERT(str == "Error: Module cannot be recursively evaluated: file:///bar");
+    });
+  });
+}
+
+// ======================================================================================
+
+KJ_TEST("Recursively require ESM from CJS required from ESM fails as expected") {
+  PREAMBLE([&](Lock& js) {
+    ResolveObserverImpl observer;
+    CompilationObserver compilationObserver;
+
+    ModuleBundle::BundleBuilder bundleBuilder;
+
+    // In this test, we have an ESM module (bar) that imports a CJS style
+    // module (foo) that synchronously tries to require the ESM module (bar).
+    // This is not allowed because the CJS module cannot successfully require
+    // a module that is still in the process of being evaluated.
+
+    bundleBuilder.addSyntheticModule("foo",
+        Module::newCjsStyleModuleHandler<TestType, TestIsolate_TypeWrapper>(
+            kj::str("exports = require('bar');"),
+            kj::str("foo")));
+
+    auto bar = kj::str("export default 123; await import('foo');");
+    bundleBuilder.addEsmModule("bar", bar.slice(0, bar.size()).attach(kj::mv(bar)));
+
+    auto registry = ModuleRegistry::Builder(observer).add(bundleBuilder.finish()).finish();
+
+    registry->attachToIsolate(js, compilationObserver);
+
+    js.tryCatch([&] {
+      ModuleRegistry::resolve(js, "file:///bar", "default"_kjc);
+      JSG_FAIL_REQUIRE(Error, "Should have failed");
+    }, [&](Value exception) {
+      auto str = kj::str(exception.getHandle(js));
+      KJ_ASSERT(str == "Error: The module evaluation did not complete synchronously. "
+                       "This is not permitted for synchronous require(...). "
+                       "Use await import(...) instead.");
+    });
+  });
+}
+
+// ======================================================================================
+
+KJ_TEST("Recursively require ESM from CJS required from ESM fails as expected (2)") {
```

Comment:
It took me a while to find the difference from the previous test... Let's rename them to e.g. `<test name> (await import)` and `<test name> (synchronous import)` to highlight the difference

Reply:
```suggestion
KJ_TEST("Recursively require ESM from CJS required from ESM fails as expected (static import)") {
```